webapp/src/app.tsx

```diff
@@ -1936,6 +1918,166 @@ function initLogin() {
+                showDownloading();
+                sendElectronMessage("update", {
+                    targetVersion: args.targetVersion
+                });
+            }
+        });
+    }
+
+    function onUpdateNotAvailable() {
+        core.confirmAsync({
+            body: lf("You are using the latest version available."),
+            header: lf("Good to go!"),
+            agreeLbl: lf("Ok"),
+            hideCancel: true
+        });
+    }
+
+    function onUpdateCheckError() {
+        displayUpdateError(lf("Unable to check for update"), lf("Ok"));
+    }
+
+    function onUpdateDownloadError(args: any) {
+        let isCritical = args && args.isCritical;
+
+        hideDownloading();
+        displayUpdateError(lf("There was an error downloading the update"), isCritical ? lf("Quit app") : lf("Ok"))
+            .finally(() => {
+                if (isCritical) {
+                    sendElectronMessage("quit");
+                }
+            });
+    }
+
+    function displayUpdateError(header: string, btnLabel: string) {
+        return core.confirmAsync({
+            header,
+            body: lf("Please ensure you are connected to the Internet and try again later."),
+            agreeClass: "red",
+            agreeIcon: "cancel",
+            agreeLbl: btnLabel,
+            hideCancel: true
+        });
+    }
+
+    pxt.log('initializing electron socket');
+    electronSocket = new WebSocket('ws://localhost:3233/' + Cloud.localToken + '/electron');
+    electronSocket.onopen = (ev) => {
+        pxt.log('electron: socket opened');
+        sendElectronMessage("ready");
+    }
+    electronSocket.onclose = (ev) => {
+        pxt.log('electron: socket closed');
+        electronSocket = null;
+    }
+    electronSocket.onmessage = (ev) => {
+        try {
+            let msg = JSON.parse(ev.data) as ElectronMessage;
+
+            switch (msg.type) {
+                case "critical-update":
+                    onCriticalUpdate(msg.args);
+                    break;
+                case "update-available":
+                    onUpdateAvailable(msg.args);
+                    break;
+                case "update-not-available":
+                    onUpdateNotAvailable();
+                    break;
+                case "update-check-error":
+                    onUpdateCheckError();
+                    break;
+                case "update-download-error":
+                    onUpdateDownloadError(msg.args);
+                    break;
+                default:
+                    pxt.debug('unknown electron message: ' + ev.data);
+                    break;
+            }
+        }
+        catch (e) {
+            pxt.debug('unknown electron message: ' + ev.data);
+        }
+    }
+}
+
+function sendElectronMessage(type: string, args?: any) {
+    if (!electronSocket) {
+        return;
+    }
+
+    let message: ElectronMessage = {
+        type: type,
+        args: args || void 0
```

Comment:
can we simply use "undefined" instead of "void 0"? TypeScript or obfuscator can do the "void 0" stuff.